contracts/lotto/src/contract.rs

```diff
@@ -0,0 +1,1144 @@
+        None => config.nois_proxy,
+    };
+
+    let lotto_nonce = lotto_nonce.unwrap_or(config.lotto_nonce);
+    let protocol_commission_percent =
+        protocol_commission_percent.unwrap_or(config.protocol_commission_percent);
+    let creator_commission_percent =
+        creator_commission_percent.unwrap_or(config.creator_commission_percent);
+
+    let is_paused = is_paused.unwrap_or(config.is_paused);
+
+    // TODO Check that the commissions are less than 100% and that the new values don't mess up with currently running lottos
+
+    let new_config = Config {
+        manager,
+        nois_proxy,
+        lotto_nonce,
+        allowlisted_recipients,
+        protocol_commission_percent,
+        creator_commission_percent,
+        is_paused,
+    };
+
+    CONFIG.save(deps.storage, &new_config)?;
+
+    Ok(Response::default().add_attribute("action", "set_config"))
+}
+
+fn execute_update_allow_listed_recipients(
+    deps: DepsMut,
+    info: MessageInfo,
+    add: Vec<String>,
+    remove: Vec<String>,
+) -> Result<Response, ContractError> {
+    let mut config = CONFIG.load(deps.storage)?;
+    ensure_eq!(info.sender, config.manager, ContractError::Unauthorized);
+
+    let mut updated_allowlist = Vec::new();
+
+    // Add new addresses to the allowlist
+    for recipient in add {
+        let addr = deps.api.addr_validate(&recipient)?;
+        if !config.allowlisted_recipients.contains(&addr) {
+            updated_allowlist.push(addr);
+        }
+    }
+
+    // Remove addresses from the allowlist
+    for recipient in &config.allowlisted_recipients {
+        let recipient_addr = deps.api.addr_validate(recipient.as_str())?;
+        if !remove.contains(&recipient_addr.to_string()) {
+            updated_allowlist.push(recipient_addr);
+        }
+    }
+
+    config.allowlisted_recipients = updated_allowlist;
+
+    CONFIG.save(deps.storage, &config)?;
+
+    Ok(Response::default().add_attribute("action", "set_config"))
+}
+
+fn execute_buy_ticket(
+    deps: DepsMut,
+    env: Env,
+    info: MessageInfo,
+    lotto_id: u64,
+) -> Result<Response, ContractError> {
+    if !LOTTOS.has(deps.storage, lotto_id) {
+        return Err(ContractError::LottoNotFound {});
+    }
+
+    let mut lotto = LOTTOS.load(deps.storage, lotto_id)?;
+    let ticket_price = lotto.clone().ticket_price;
+
+    // Not sure the best way to go about validating the coin
+    validate_payment(&ticket_price, info.funds.as_slice())?;
+
+    // Check if lotto is active
+    if env.block.time >= lotto.expiration {
+        return Err(ContractError::LottoDepositStageEnded {});
+    }
+    // Increment total deposit
+    let balance: Coin = info
+        .funds
+        .iter()
+        .filter(|coin| coin.denom == ticket_price.denom)
+        .last()
+        .unwrap()
+        .clone();
+
+    lotto.balance += balance.amount;
+    // Add participant address
+    lotto.participants.push(info.clone().sender);
```

Comment:
```suggestion
    lotto.participants.push(info.sender.clone());
```

No need to clone the whole struct when you just need the sender